martin/src/cog/mod.rs

```diff
@@ -237,67 +244,78 @@ impl SourceConfigExtras for CogConfig {
-        tiff::ColorType::RGB(8) | tiff::ColorType::RGBA(8)
-    ) {
-        Err(CogError::NotSupportedColorTypeAndBitDepth(
-            color_type,
-            path.clone(),
-        ))?;
+        Err(CogError::NotSupportedChunkType(path.to_path_buf()))?;
     }
 
+    // see https://docs.ogc.org/is/21-026/21-026.html#_planar_configuration_considerations and https://www.verypdf.com/document/tiff6/pg_0038.htm
+    // we might support planar configuration 2 in the future
     decoder
         .get_tag_unsigned(Tag::PlanarConfiguration)
         .map_err(|e| {
-            CogError::TagsNotFound(e, vec![Tag::PlanarConfiguration.to_u16()], 0, path.clone())
+            CogError::TagsNotFound(
+                e,
+                vec![Tag::PlanarConfiguration.to_u16()],
+                0,
+                path.to_path_buf(),
+            )
         })
         .and_then(|config| {
             if config == 1 {
                 Ok(())
             } else {
                 Err(CogError::PlanarConfigurationNotSupported(
-                    path.clone(),
+                    path.to_path_buf(),
                     0,
                     config,
                 ))
             }
         })?;
 
-    let tag = decoder.get_tag_ascii_string(GdalNodata);
-    let nodata: Option<f64> = if let Ok(nodata_tag) = tag {
-        nodata_tag.parse().ok()
-    } else {
-        None
+    let color_type = decoder
+        .colortype()
+        .map_err(|e| CogError::InvalidTiffFile(e, path.to_path_buf()))?;
+
+    if !matches!(
+        color_type,
+        tiff::ColorType::RGB(8) | tiff::ColorType::RGBA(8)
+    ) {
+        Err(CogError::NotSupportedColorTypeAndBitDepth(
+            color_type,
+            path.to_path_buf(),
+        ))?;
     };
-    let images_ifd = get_images_ifd(&mut decoder);
+    Ok(())
+}
+
+#[allow(clippy::cast_possible_truncation)]
+fn get_meta(path: &PathBuf) -> Result<Meta, FileError> {
+    let tif_file = File::open(path).map_err(|e| FileError::IoError(e, path.clone()))?;
+    let mut decoder = Decoder::new(tif_file)
+        .map_err(|e| CogError::InvalidTiffFile(e, path.clone()))?
+        .with_limits(tiff::decoder::Limits::unlimited());
 
+    verify_requirments(&mut decoder, path)?;
     let mut zoom_and_ifd: HashMap<u8, usize> = HashMap::new();
     let mut zoom_and_tile_across_down: HashMap<u8, (u32, u32)> = HashMap::new();
 
-    for image_ifd in &images_ifd {
+    let nodata: Option<f64> = if let Ok(no_data) = decoder.get_tag_ascii_string(GdalNodata) {
+        no_data.parse().ok()
+    } else {
+        None
+    };
+
+    let images_ifd = get_images_ifd(&mut decoder, path);
+
+    for (idx, image_ifd) in images_ifd.iter().enumerate() {
         decoder
             .seek_to_image(*image_ifd)
             .map_err(|e| CogError::IfdSeekFailed(e, *image_ifd, path.clone()))?;
 
-        let zoom = u8::try_from(images_ifd.len() - (image_ifd + 1))
+        let zoom = u8::try_from(images_ifd.len() - (idx + 1))
```

Comment:
The result of `images_ifd.len() - (image_ifd + 1)` would be wrong when there are masks.  Updated this to `images_ifd.len() - (idx + 1)`